src/exception.c

```diff
@@ -253,7 +265,57 @@ static const char* __get_exception_name(exception_code_t code)
 		"Reserved",									// 31
 	};
 
-	return exceptionMap[code];
+
+	// When possible, by peeking into the exception state and COP0 registers
+	// we can provide a more detailed exception name.
+	uint32_t epc = ex->regs->epc + (ex->regs->cr & C0_CAUSE_BD ? 4 : 0);
+	uint32_t badvaddr = C0_BADVADDR();
+
+	switch (ex->code) {
+	case EXCEPTION_CODE_FLOATING_POINT:
+		if (ex->regs->fc31 & C1_CAUSE_DIV_BY_0) {
+			return "Floating point divide by zero";
+		} else if (ex->regs->fc31 & C1_CAUSE_INVALID_OP) {
+			return "Floating point invalid operation";
+		} else if (ex->regs->fc31 & C1_CAUSE_OVERFLOW) {
+			return "Floating point overflow";
+		} else if (ex->regs->fc31 & C1_CAUSE_UNDERFLOW) {
+			return "Floating point underflow";
+		} else if (ex->regs->fc31 & C1_CAUSE_INEXACT_OP) {
+			return "Floating point inexact operation";
+		} else {
+			return "Generic floating point";
+		}
+	case EXCEPTION_CODE_TLB_LOAD_I_MISS:
+		if (epc == badvaddr) {
+			return "Invalid program counter address";
+		} else if (badvaddr < 128) {
```

Comment:
Why does it check `< 128` for null reference? (same below)

Reply:
it's just a heuristic to catch null pointers. One common case is that you have a NULL pointer to struct, and deference one of the fields of the struct. In that case you don't get a TLB miss on exactly 0x00000000, but the address is basically the offset of the field in the struct. If the struct is larger you might instead get `Read from invalid memory address`, which is not so bad too, so I guess that's an acceptable compromise.

Reply:
Maybe we can incorporate "struct" in the message when it is not exactly zero to further help?

Reply:
Haven't thought of the array, right. I don't think it is particularly confusing. Just adding a comment will clear things for readers for sure.

Reply:
done